### PR TITLE
Implement new Face Recognition Module

### LINKED ISSUE #294: [Feature]: Grouping of images based on faces found in the images.
### Feature detail

Google Photos and iPhone Photos have a feature where photos are organized based on faces found in the image.

Also, there are some self hosted solution providers which also have such feature.
1. https://github.com/photoview/photoview
2. https://github.com/LibrePhotos/librephotos
3. https://github.com/photonixapp/photonix

It would be great if you can provide such feature in immich.

### Platform

Mobile App

### PR DESCRIPTION
**Pull request by LackesLab**
Dear Contributors, 
I like to implement a new module handling the face detection and recognition for Immich. Please find my proposal as first part in the changed README and give me feedback.
Please let me know if you have any further questions.

Best  Regards
Lukas

### CODE REVIEW

**Comment by zackpollard** [comment]
Hey, thanks for your proposal, this sounds great to me and something we would all love to see added to Immich. If you want an easier way to continue any discussion, please go ahead and join the discord here https://discord.gg/k88wPj5A

**Comment by zackpollard** [comment]
This proposal sounds great to me and something I am sure we would all love to see added into Immich!

**Comment by bertmelis** [comment]
Immich already has a DB in the stack. Wouldn't you rather want to plug into that one instead of setting up a new one?

**Comment by LackesLab** [comment]
As discussed on discord, I'll first go with a file based datastore.

**Comment by zoodyy** [comment]
I'd gladly see ML based face recognition in Immich!
To find similar face embeddings quickly (to classify a new face), use an approximate nearest neighbor library like hnswlib.

**Comment by mike-lloyd03** [comment]
The workflow looks good. There's a good chance that the algorithm will detect the same person as different faces and assign them different IDs. Can we include a workflow for merging two face IDs into one?

This would probably also be the same workflow for if a photo is misidentified as a different pre-existing face ID.

**Comment by bo0tzz** [comment]
> Can we include a workflow for merging two face IDs into one?

The plan is to apply tags to images listing things like which faces are in them, as well as other information like recognized objects or user-created tags. That functionality is intended to then include support for merging tags as well.

**Comment by eschwim** [comment]
Any chance that we could get Edge TPU (e.g. Coral USB accelerator) support for face recognition baked in? I've got a photo lib with something like 60k photos, but the 1U server I host it off of has a fairly anemic EPYC CPU with no GPU installed.    

Having tried to import my lib into both LibrePhotos and PhotoPrism (both of which have facial detection, but neither of which has TPU support) has melted my server down a few times =D   I already have a Coral TPU installed (for Frigate), so I would be happen to test.

**Comment by alextran1502** [comment]
I would love to test it but I don't have a hardware at the moment

**Comment by bo0tzz** [comment]
I looked into TPU support a little bit. The problem is that Coral TPUs are Tensorflow devices, while we're using pytorch, so there would have to be a step that transpiles the models. That would complicate things quite a bit, I think.

File: machine-learning/src/main.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
     model="hustvl/yolos-tiny"
 )
 
+face_processor = FaceProcessor()
```
Comment (jrasm91) [inline machine-learning/src/main.py:19]: Do we want to expose any options here via env variables?

File: machine-learning/src/main.py [hunk 2]
```diff
@@ -38,6 +39,11 @@ def image_classification():
     assetPath = request.json['thumbnailPath']
     return run_engine(classifier, assetPath), 201
 
+@server.route("/facial-recognition/recognize-persons", methods=['POST'])
+def facial_recognition():
+    assetPath = request.json['thumbnailPath']
+    return face_processor.process_image(assetPath), 201
+    #return run_engine(classifier, assetPath), 201
```
Comment (jrasm91) [inline machine-learning/src/main.py:46]: This can be removed, yeah?

**Review by jrasm91 — APPROVED** [review]
Looks good! Would it be best to merge this as is and work on face management in a separate PR?

**Comment by jagjordi** [comment]
> I would love to test it but I don't have a hardware at the moment

@alextran1502 they are in stock now on seeedstudio! I got two today after waiting for several months
https://www.seeedstudio.com/Coral-USB-Accelerator-p-2899.html

**Comment by jagjordi** [comment]
I am also available to test edge TPU related releases

**Comment by bo0tzz** [comment]
Just a heads up, before people start buying Corals for use with Immich: Our ML system is currently using pytorch, and the Coral is only compatible with Tensorflow.

**Comment by jagjordi** [comment]
Most people in the self-hosted community have a Coral for use with frigate, so I would say that it would be good eventually if the ML system is ported to Tensorflow (coral actually only supports tensorflow lite)

**Comment by mike-lloyd03** [comment]
I too have a Coral for Frigate and would love to use it for more services. But now that I think about it, I'm not sure how big of a difference it will make with Immich. It'll probably make object detection run faster on initial import of a large batch of images. But as I upload 5 or 6 images a day, I doubt the CPU load will be all that significant.

**Comment by jagjordi** [comment]
I agree that daily normal usage is not very big but the initial import is a significant part of the user experience of the system. I have around 300k images in my library and with the current setup I'm getting less than 1k images per hour of performance of tagging. That means it would take something like 4 months to run the initial tagging. And that's with the CPU at 100% for 4 months! Imagine the power consumption!

In short, I would not dismiss this feature just because "it's only useful for the initial import" sins for many users the initial import is a big part of the system.

**Comment by alextran1502** [comment]
I believe SBERT, the technology for CLIP search uses Pytorch, that is why we are having Pytorch as dependency.

**Comment by normanu** [comment]
Face recognition would be awesome!!!
Is this something I can help with by testing it?
If so, how can I test :)

**Comment by jrasm91** [comment]
@LackesLab has finished the initial python implementation for facial recognition. However, there is still a lot to do with regards to clustering, labeling, merging, and other face/person related management tasks, which we're planning on building into immich itself. For example,

- Faces metadata to be saved in `asset.smartInfo`
- Face embeddings to be to pushed to `typesense`, a vector database, to enable similarity distance queries
- Auto-link/associate faces to people
- Track people, with abilities to label, merge, manage, link to a user, etc.
- Generate and serve optimized face images (ideally from the original images, not the resized thumbnail)
- Integrate faces into the web/mobile app
- Search by people/faces
- New admin job/queue for tracking and regenerating face embeddings
- etc.

To make it easier for the core immich dev team to help contribute to the remaining tasks, we've recreated this PR in the immich repo itself as #2180.
